service/importer/definition.go

```diff
@@ -1,108 +1,111 @@
 
 	// Description is service description.
-	Description string `yaml:"description" json:"description,omitempty"`
+	Description string `yaml:"description" json:"description,omitempty" validate:"printascii"`
 
 	// Tasks are the list of tasks that service can execute.
-	Tasks map[string]*Task `yaml:"tasks" json:"tasks,omitempty"`
+	Tasks map[string]*Task `yaml:"tasks" json:"tasks,omitempty" validate:"dive,keys,printascii,endkeys,required"`
 
 	// Events are the list of events that service can emit.
-	Events map[string]*Event `yaml:"events" json:"events,omitempty"`
+	Events map[string]*Event `yaml:"events" json:"events,omitempty" validate:"dive,keys,printascii,endkeys,required"`
 
 	// Dependencies are the Docker containers that service can depend on.
-	Dependencies map[string]*Dependency `yaml:"dependencies" json:"dependencies,omitempty"`
+	Dependencies map[string]*Dependency `yaml:"dependencies" json:"dependencies,omitempty" validate:"dive,keys,printascii,ne=service,endkeys,required"`
 
 	// Configuration is the Docker container that service runs inside.
 	Configuration *Dependency `yaml:"configuration" json:"configuration,omitempty"`
 
 	// Repository holds the service's repository url if it's living on
 	// a Git host.
-	Repository string `yaml:"repository" json:"repository,omitempty"`
+	Repository string `yaml:"repository" json:"repository,omitempty" validate:"omitempty,uri"`
 }
 
 // Event describes a service task.
 type Event struct {
 	// Name is the name of event.
-	Name string `yaml:"name" json:"name,omitempty"`
+	Name string `yaml:"name" json:"name,omitempty" validate:"printascii"`
 
 	// Description is the description of event.
-	Description string `yaml:"description" json:"description,omitempty"`
+	Description string `yaml:"description" json:"description,omitempty" validate:"printascii"`
 
 	// Data holds the input parameters of event.
-	Data map[string]*Parameter `yaml:"data" json:"data,omitempty"`
+	Data map[string]*Parameter `yaml:"data" json:"data,omitempty" validate:"required,dive,keys,printascii,endkeys,required"`
 }
 
 // Dependency represents a Docker container and it holds instructions about
 // how it should run.
 type Dependency struct {
 	// Image is the Docker image.
-	Image string `yaml:"image" json:"image,omitempty"`
+	Image string `yaml:"image" json:"image,omitempty" validate:"printascii"`
 
 	// Volumes are the Docker volumes.
-	Volumes []string `yaml:"volumes" json:"volumes,omitempty"`
+	Volumes []string `yaml:"volumes" json:"volumes,omitempty" validate:"unique,dive,printascii"`
 
 	// VolumesFrom are the docker volumes-from from.
-	VolumesFrom []string `yaml:"volumesfrom" json:"volumefrom,omitempty"`
+	VolumesFrom []string `yaml:"volumesfrom" json:"volumefrom,omitempty" validate:"unique,dive,printascii"`
 
 	// Ports holds ports configuration for container.
-	Ports []string `yaml:"ports" json:"ports,omitempty"`
+	Ports []string `yaml:"ports" json:"ports,omitempty" validate:"unique,dive,numeric"`
```

Comment:
need a validate range 0-65535 (but it cannot be done with validate it has to be done in func)

Reply:
So it's from different vliadator package, function is port looks like :

```
// IsPort checks if a string represents a valid port
func IsPort(str string) bool {
	if i, err := strconv.Atoi(str); err == nil && i > 0 && i < 65536 {
		return true
	}
	return false
}
```

And we could add this as our custom validator to package that we are using